### PR TITLE
[feat] add BaseShop abstract class and refactor CandyShop types

### PR DESCRIPTION
**Pull request by beepsaidblue**

### CODE REVIEW

File: core/sdk/src/shop/base/BaseShop.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export abstract class BaseShop {
+  constructor(params: any) {}
+
+  // Properties
+  abstract get candyShopAddress(): string;
+  abstract get candyShopCreatorAddress(): string;
+  abstract get treasuryMint(): string;
+  abstract get settings(): Partial<ShopSettings>;
+  abstract get blockchain(): Blockchain;
+  abstract get baseUnitsPerCurrency(): number;
```
Comment (RustySol) [inline core/sdk/src/shop/base/BaseShop.ts:12]: this should be the same as `currencyDecimals`, I think can remove this one at some point

**Review by RustySol — APPROVED** [review]

File: core/sdk/src/shop/base/BaseShop.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  abstract cancelAuction(params: any): Promise<string>;
+  abstract bidAuction(params: any): Promise<string>;
+  abstract withdrawAuctionBid(params: any): Promise<string>;
+  abstract buyNowAuction(params: any): Promise<string>;
+  abstract settleAndDistributeAuctionProceeds(params: any): Promise<string>;
+
+  // Edition Drop
+  abstract commitMasterNft(params: any): Promise<string>;
+  abstract mintNewPrint(params: any): Promise<string>;
+  abstract redeemDrop(params: any): Promise<string>;
```
Comment (haVincy) [inline core/sdk/src/shop/base/BaseShop.ts:34]: All abstract methods mean the inheritor need to implement its own stuffs, however, I wonder if there's method will have common setup before doing anything? If in this case, we can put common actions in concrete method and let inheritor to call `super.xxx` to do the common action before performing different.

File: core/sdk/src/shop/base/BaseShop.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  abstract get candyShopAddress(): string;
+  abstract get candyShopCreatorAddress(): string;
+  abstract get treasuryMint(): string;
+  abstract get settings(): Partial<ShopSettings>;
+  abstract get blockchain(): Blockchain;
+  abstract get baseUnitsPerCurrency(): number;
+  abstract get currencySymbol(): string;
+  abstract get currencyDecimals(): number;
+  abstract get priceDecimalsMin(): number;
+  abstract get priceDecimals(): number;
```
Comment (haVincy) [inline core/sdk/src/shop/base/BaseShop.ts:16]: I feel most of these properties are common fields? Don't have to be `abstract`, can define the `private` members of these properties in `BaseShop` itself and return it from here by get methods.


```ts
export abstract class BaseShop {
  private _candyShopAddress: web3.PublicKey;
  private _candyShopCreatorAddress: web3.PublicKey;
  private _treasuryMint: web3.PublicKey;
  private _settings: CandyShopSettings;
  private _baseUnitsPerCurrency: number;
  ...

  constructor(params: CandyShopConstructorParams) {
       this. _candyShopAddress = params.candyShopAddress;
       this._candyShopCreatorAddress = params.candyShopCreatorAddress;
       this._treasuryMint = params.treauryMint;
       ...
  }

  get candyShopAddress(): string {
     return this._candyShopAddress;
  }  
  ....
}

export CandyShop extends BaseShop {
    constructor(params: CandyShopConstructorParams, SolParams: CandyShopSolParams) {
       super(params);
       ...
       // assign those Sol specific construct params
      ....
  }
}

// Usage

const candyShop = new CandyShop(candyShopConstructorParams, candyShopSolParams);
console.log(candyShop.candyShopAddress())

```

**Review by haVincy — APPROVED** [review]
LGTM. some details can be done in near future. Need to resolve the conflicts with TOT master.

**Pull request merged**